flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskStateManagerImpl.java

```diff
@@ -68,18 +70,36 @@
 	/** The checkpoint responder through which this manager can report to the job manager. */
 	private final CheckpointResponder checkpointResponder;
 
+	private final ChannelStateReader channelStateReader;
+
 	public TaskStateManagerImpl(
-		@Nonnull JobID jobId,
-		@Nonnull ExecutionAttemptID executionAttemptID,
-		@Nonnull TaskLocalStateStore localStateStore,
-		@Nullable JobManagerTaskRestore jobManagerTaskRestore,
-		@Nonnull CheckpointResponder checkpointResponder) {
+			@Nonnull JobID jobId,
+			@Nonnull ExecutionAttemptID executionAttemptID,
+			@Nonnull TaskLocalStateStore localStateStore,
+			@Nullable JobManagerTaskRestore jobManagerTaskRestore,
+			@Nonnull CheckpointResponder checkpointResponder) {
+		this(jobId,
+			executionAttemptID,
+			localStateStore,
+			jobManagerTaskRestore,
+			checkpointResponder,
+			new ChannelStateReaderImpl(jobManagerTaskRestore == null ? new TaskStateSnapshot() : jobManagerTaskRestore.getTaskStateSnapshot())
+		);
+	}
 
-		this.jobId = jobId;
+	TaskStateManagerImpl(
```

Comment:
only need private atm

Reply:
It could be used in tests.

Reply:
@VisibleForTesting

Reply:
There is nothing wrong with using this constructor in production code too.

Reply:
Yes, I agree with your point to some extent. I think the key concern is how we define the access modifier based on two considerations.

- Based on current demands: if so, it should be private ATM, and then further extend it  by demands if necessary future.

- Based on future considerations: if so, it can be defined as package public now, even public.  But it is hard to say whether it is alway fitting the expectation, then it might seem unnecessary for long time. If taking this option, we might even remove `@VisibleForTesting` annotation for previous usages, because any constructors might have the possibility to be used in core codes future.

Anyway, I can accept both options and do not think it is a big issue. Just share some thoughts, feel free to take for your favor. :)